### PR TITLE
Avoid `numpy.can_cast` call to improve guess routine 

### PR DESCRIPTION
**Pull request by okuta**
`numpy.can_cast` is called many times.
The new routine uses ` numpy.min_scalar_type` instead of `numpy.can_cast`.

I added benchmark result.

### CODE REVIEW

**Comment by okuta** [comment]
Jenkins, test this please

File: cupy/core/_kernel.pyx [hunk 1]
```diff
@@ -701,28 +706,42 @@ cdef inline bint _check_should_use_min_scalar(list in_args) except? -1:
...
 
 
-cdef tuple _guess_routine(name, dict cache, list ops, list in_args, dtype):
+cdef dict _mst_unsigned_to_signed = {
+    i: (numpy.iinfo(j).max, (i, j))
+    for i, j in [(numpy.dtype(i).type, numpy.dtype(i.lower()).type)
+                 for i in "BHILQ"]}
+cdef _numpy_min_scalar_type = numpy.min_scalar_type
+
+cdef _min_scalar_type(x):
```
Comment (toslunar) [inline cupy/core/_kernel.pyx]: Could you add a comment to the function?  For example,
"A non-negative integer may have two locally minimum scalar types: signed/unsigned integer. Return both for `can_cast`, while `numpy.min_scalar_type` only returns the unsigned type."

**Comment by toslunar** [comment]
No fixes are needed between (u)int types and float types because of the surprising behavior of NumPy:
```
>>> np.can_cast(256, np.float16)
False
>>> np.arange(3, dtype=np.float16) * 256
array([  0., 256., 512.], dtype=float32)
>>> np.can_cast(256.0, np.float16)
True
>>> np.arange(3, dtype=np.float16) * 256.0
array([  0., 256., 512.], dtype=float16)
```

File: cupy/core/_kernel.pyx (hunk 1 above)
Comment (okuta) [reply cupy/core/_kernel.pyx]: Thanks

**Comment by okuta** [comment]
I tested with NumPy 1.17.4 .
```
% python -m timeit -s "import cupy; a = cupy.zeros((16,), 'f'); out = a.copy()" "cupy.add(a, 1, out=out)"
```

before with NumPy 17.4
```
10000 loops, best of 3: 38.1 usec per loop
```


after with NumPy 17.4
```
10000 loops, best of 3: 21.3 usec per loop
```

test with NumPy 16.5
```
10000 loops, best of 3: 17.8 usec per loop
```

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**